### PR TITLE
Modular alerts

### PR DESCRIPTION
**Pull request by IDoneShaveIt**
Using `alert_fields` field under test meta to configure the alert of the test.
If there is no alert_fields, we return all of the alert's fields

### CODE REVIEW

**Comment by Maayan-s** [comment]
@IDoneShaveIt 
This feature requires docs

File: elementary/monitor/alerts/model.py [hunk 1]
```diff
@@ -40,73 +40,125 @@ def to_slack(self, is_slack_workflow: bool = False) -> SlackMessageSchema:
...
-        icon = ":small_red_triangle:"
-        if self.status == "warn":
-            icon = ":warning:"
-        slack_message = {"attachments": [{"blocks": []}]}
-        self._add_text_section_to_slack_msg(slack_message, f"{icon} *dbt model alert*")
-        self._add_fields_section_to_slack_msg(
-            slack_message,
+        tags = prettify_json_str_set(self.tags)
+        owners = prettify_json_str_set(self.owners)
+        subscribers = prettify_json_str_set(", ".join(self.subscribers))
```
Comment (oravi) [inline elementary/monitor/alerts/model.py:45]: Why do we handle subscribers differently than owners and tags?

File: elementary/monitor/alerts/alert.py [hunk 2]
```diff
@@ -56,50 +63,101 @@ def __init__(
...
-            return section_msg
-        return (
-            section_msg[
-                : SectionBlock.text_max_length
-                - len(cls._CONTINUATION_SYMBOL)
-                - cls._LONGEST_MARKDOWN_SUFFIX_LEN
-            ]
-            + cls._CONTINUATION_SYMBOL
-            + section_msg[-cls._LONGEST_MARKDOWN_SUFFIX_LEN :]
+
```
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply elementary/monitor/alerts/alert.py:66]: Its a very specific exception which is only relevant for alerts slack message so it doesn't feel right to me to put it on a more generic exceptions file.

File: elementary/monitor/alerts/alert.py [hunk 3]
```diff
@@ -56,50 +63,101 @@ def __init__(
...
+        super().__init__()
+
+    def get_slack_message(
+        self,
+        title: Optional[SlackBlocksType] = None,
+        preview: Optional[SlackBlocksType] = None,
+        result: Optional[SlackBlocksType] = None,
+        configuration: Optional[SlackBlocksType] = None,
+    ) -> SlackMessageSchema:
+        alert = AlertSlackMessageSchema(
```
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply elementary/monitor/alerts/alert.py:92]: We can skip this schema and just pass the parts, but I like the Idea that we have a schema that validates that alerts are always built from title, preview and details.
I think that using schema for typing is also a good advantage and helps a lot to understand what data passes and how it looks - using dict, list and basic typing doesn't really provides you the real structure of those "more complex" vars, and when you pass dict, there is no contract for the data it contains, while schema does.
WDYT?

File: elementary/clients/slack/schema.py [hunk 4]
```diff
@@ -1,7 +1,10 @@
-from typing import Optional
+from typing import List, Optional
 
 from pydantic import BaseModel
 
+SlackBlockType = dict
+SlackBlocksType = List[Optional[SlackBlockType]]
```
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply elementary/clients/slack/schema.py:6]: In my opinion yes 🙂.
dict is generic, and while my typing can be `preview: List[dict]` when I use `preview: List[SlackBlockType]` I always know that the dict I will see will represent a list of Slack blocks.
IMO it makes developing easier and clearer

File: elementary/monitor/alerts/alert.py [hunk 5]
```diff
@@ -56,50 +63,101 @@ def __init__(
...
+    ) -> None:
+        self.preview_blocks = preview_blocks
+        self.message = message
+        super().__init__(self.message)
+
+    def __str__(self) -> str:
+        return f"{len(self.preview_blocks)} provieded -> {self.message}"
+
+
+class AlertSlackMessageBuilder(SlackMessageBuilder):
```
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply elementary/monitor/alerts/alert.py:81]: slack_message_builder is the base class that creates the base slack blocks. It is relevant for all slack messages what we will use (including report summary message for example).
The alert slack message builder is the builder for alerts which include the specific logic for alert messages.
You can think of it like an sdk, and class that wraps the sdk for a specific use.
I merge the AlertSlackMessageBuilder into the Alert class  though

File: elementary/monitor/alerts/alert.py [hunk 6]
```diff
@@ -56,50 +64,110 @@ def __init__(
...
+    @staticmethod
+    def prettify_list_variations(list_variation: Union[List[str], str]) -> str:
+        if isinstance(list_variation, str):
+            return prettify_json_str_set(list_variation)
+
+        elif isinstance(list_variation, list):
+            return ", ".join(set(list_variation))
+
+        else:
+            return ""
```
Comment (oravi) [inline elementary/monitor/alerts/alert.py:173]: I'd return the list_variation in this case

File: elementary/monitor/alerts/alert.py [hunk 7]
```diff
@@ -56,50 +64,110 @@ def __init__(
...
+            while blocks_counter < MAX_ALERT_PREVIEW_BLOCKS:
+                padded_preview_blocks.append(cls.create_empty_section_block())
+                blocks_counter += 1
+            return padded_preview_blocks
+
+        else:
+            return preview_blocks
+
+    @staticmethod
+    def prettify_list_variations(list_variation: Union[List[str], str]) -> str:
```
Comment (oravi) [inline elementary/monitor/alerts/alert.py:165]: cosmetics - I'd rename to just prettify_list

**Pull request merged**